File: packages/python/src/erpc_sdk/rest.py
"""Typed account, usage, price, SSE, and Cloud read clients."""

from __future__ import annotations

import asyncio
import json
import math
import re
from collections.abc import AsyncIterator, Mapping, Sequence
from datetime import datetime
from typing import cast
from urllib.parse import quote

from .errors import ErpcAbortedError, ErpcConfigError, ErpcInvalidResponseError
from .transport import QueryValue, RestTransport
from .types import (
    DEFAULT_REQUEST_OPTIONS,
    CloudCredit,
    CloudOffering,
    CloudResource,
    CloudResourceStatus,
    JsonValue,
    MonthlyApiKeyUsage,
    PriceFeedMetadata,
    PriceStreamEvent,
    PriceUpdateResponse,
    RequestOptions,
    TokenBalance,
)

_YEAR_MONTH = re.compile(r"^\d{4}-(0[1-9]|1[0-2])$")
_RESOURCE_KINDS = {"bare-metal", "solana-grpc", "solana-shredstream", "vps"}
_RESOURCE_MODES = {"dedicated", "direct", "shared"}
_BILLING_STATUSES = {"active", "grace-period", "inactive", "suspended"}
_ALERT_LEVELS = {"critical", "normal", "suspended", "warning"}


def _mapping(value: object) -> dict[str, object]:
    if not isinstance(value, dict) or not all(isinstance(key, str) for key in value):
        raise ErpcInvalidResponseError()
    return cast(dict[str, object], value)


def _string(value: object) -> str:
    if not isinstance(value, str):
        raise ErpcInvalidResponseError()
    return value


def _number(value: object, *, integer: bool = False, nonnegative: bool = False) -> int | float:
    if not isinstance(value, (int, float)) or isinstance(value, bool) or not math.isfinite(value):
        raise ErpcInvalidResponseError()
    if integer and not isinstance(value, int):
        raise ErpcInvalidResponseError()
    if nonnegative and value < 0:
        raise ErpcInvalidResponseError()
    return value


def _datetime(value: object) -> str:
    text = _string(value)
    try:
        datetime.fromisoformat(text.replace("Z", "+00:00"))
    except ValueError as error:
        raise ErpcInvalidResponseError() from error
    return text


class AccountClient:
    def __init__(self, transport: RestTransport) -> None:
        self._transport = transport

    async def get_token_balance(
        self, options: RequestOptions = DEFAULT_REQUEST_OPTIONS
    ) -> TokenBalance:
        value = await self._transport.get("/v3/erpc/token-balance", options=options)
        data = _mapping(value)
        if data.get("plan") not in {"business", "developer", "free", "pro"}:
            raise ErpcInvalidResponseError("ERPC returned an invalid token balance")
        _number(data.get("max_tokens"), integer=True)
        _number(data.get("remaining_tokens"), integer=True)
        if data.get("next_refill_at") is not None:
            _string(data.get("next_refill_at"))
        return cast(TokenBalance, data)


class UsageClient:
    def __init__(self, transport: RestTransport) -> None:
        self._transport = transport

    async def get_monthly_api_key_usage(
        self,
        year_month: str | None = None,
        options: RequestOptions = DEFAULT_REQUEST_OPTIONS,
    ) -> MonthlyApiKeyUsage:
        if year_month is not None and not _YEAR_MONTH.fullmatch(year_month):
            raise ErpcConfigError("year_month must use YYYY-MM format")
        value = await self._transport.get(
            "/v3/user/api-keys/usage", {"yearMonth": year_month}, options
        )
        envelope = _mapping(value)
        if envelope.get("success") is not True:
            raise ErpcInvalidResponseError(
                "ERPC returned an invalid monthly API key usage response"
            )
        usage = _mapping(envelope.get("message"))
        try:
            year = _string(usage.get("yearMonth"))
            if not _YEAR_MONTH.fullmatch(year):
                raise ErpcInvalidResponseError()
            for field in ("keyCount", "totalCount", "totalCredits"):
                _number(usage.get(field))
            if not isinstance(usage.get("hasStrandedUsage"), bool):
                raise ErpcInvalidResponseError()
            if usage.get("updatedAt") is not None:
                _string(usage.get("updatedAt"))
            self._validate_chains(usage.get("chains"))
            api_keys = usage.get("apiKeys")
            if not isinstance(api_keys, list):
                raise ErpcInvalidResponseError()
            for value in api_keys:
                item = _mapping(value)
                _string(item.get("apiKeyLast4"))
                for field in ("apiKeyLength", "count", "credits"):
                    _number(item.get(field))
                if item.get("keyId") is not None:
                    _number(item.get("keyId"))
                if item.get("updatedAt") is not None:
                    _string(item.get("updatedAt"))
                self._validate_chains(item.get("chains"))
        except ErpcInvalidResponseError as error:
            raise ErpcInvalidResponseError(
                "ERPC returned an invalid monthly API key usage response"
            ) from error
        return cast(MonthlyApiKeyUsage, usage)

    @staticmethod
    def _validate_chains(value: object) -> None:
        if not isinstance(value, list):
            raise ErpcInvalidResponseError()
        for raw_chain in value:
            chain = _mapping(raw_chain)
            _string(chain.get("chain"))
            _number(chain.get("count"))
            _number(chain.get("credits"))
            if chain.get("updatedAt") is not None:
                _string(chain.get("updatedAt"))
            methods = chain.get("methods")
            if not isinstance(methods, list):
                raise ErpcInvalidResponseError()
            for raw_method in methods:
                method = _mapping(raw_method)
                _string(method.get("method"))
                for field in ("count", "creditCost", "credits"):
                    _number(method.get(field))
                if method.get("updatedAt") is not None:
                    _string(method.get("updatedAt"))


class PriceClient:
    def __init__(self, transport: RestTransport) -> None:
        self._transport = transport

    async def get_price_feeds(
        self,
        *,
        asset_type: str | None = None,
        query: str | None = None,
        options: RequestOptions = DEFAULT_REQUEST_OPTIONS,
    ) -> list[PriceFeedMetadata]:
        value = await self._transport.get(
            "/v2/price_feeds", {"query": query, "asset_type": asset_type}, options
        )
        if not isinstance(value, list):
            raise ErpcInvalidResponseError("ERPC returned invalid price feed metadata")
        for item in value:
            metadata = _mapping(item)
            _string(metadata.get("id"))
            attributes = metadata.get("attributes")
            if attributes is not None and (
                not isinstance(attributes, dict)
                or not all(
                    isinstance(key, str) and isinstance(entry, str)
                    for key, entry in attributes.items()
                )
            ):
                raise ErpcInvalidResponseError("ERPC returned invalid price feed metadata")
        return cast(list[PriceFeedMetadata], value)

    async def get_latest_price_updates(
        self,
        ids: Sequence[str],
        *,
        encoding: str | None = None,
        parsed: bool | None = None,
        ignore_invalid_price_ids: bool | None = None,
        options: RequestOptions = DEFAULT_REQUEST_OPTIONS,
    ) -> PriceUpdateResponse:
        value = await self._transport.get(
            "/v2/updates/price/latest",
            self._update_query(ids, encoding, parsed, ignore_invalid_price_ids),
            options,
        )
        return self._price_update(value)

    async def get_price_updates_at_timestamp(
        self,
        publish_time: int | str,
        ids: Sequence[str],
        *,
        encoding: str | None = None,
        parsed: bool | None = None,
        ignore_invalid_price_ids: bool | None = None,
        options: RequestOptions = DEFAULT_REQUEST_OPTIONS,
    ) -> PriceUpdateResponse:
        value = await self._transport.get(
            f"/v2/updates/price/{quote(str(publish_time), safe='')}",
            self._update_query(ids, encoding, parsed, ignore_invalid_price_ids),
            options,
        )
        return self._price_update(value)

    async def get_latest_publisher_stake_caps(
        self,
        *,
        encoding: str | None = None,
        parsed: bool | None = None,
        options: RequestOptions = DEFAULT_REQUEST_OPTIONS,
    ) -> dict[str, JsonValue]:
        value = await self._transport.get(
            "/v2/updates/publisher_stake_caps/latest",
            {"encoding": encoding, "parsed": parsed},
            options,
        )
        return cast(dict[str, JsonValue], _mapping(value))

    async def stream_price_updates(
        self,
        ids: Sequence[str],
        *,
        encoding: str | None = None,
        parsed: bool | None = None,
        ignore_invalid_price_ids: bool | None = None,
        allow_unordered: bool | None = None,
        benchmarks_only: bool | None = None,
        options: RequestOptions = DEFAULT_REQUEST_OPTIONS,
    ) -> AsyncIterator[PriceStreamEvent]:
        query = {
            **self._update_query(ids, encoding, parsed, ignore_invalid_price_ids),
            "allow_unordered": allow_unordered,
            "benchmarks_only": benchmarks_only,
        }
        async with self._transport.stream("/v2/updates/price/stream", query, options) as response:
            buffer = ""
            iterator = response.aiter_text().__aiter__()
            while True:
                try:
                    chunk = await self._next_chunk(iterator, options.cancel_event)
                except StopAsyncIteration:
                    break
                buffer += chunk
                while match := re.search(r"\r?\n\r?\n", buffer):
                    block, buffer = buffer[: match.start()], buffer[match.end() :]
                    event = self._parse_sse_block(block)
                    if event is not None:
                        yield event
            event = self._parse_sse_block(buffer)
            if event is not None:
                yield event

    @staticmethod
    async def _next_chunk(iterator: AsyncIterator[str], cancel_event: asyncio.Event | None) -> str:
        chunk_task: asyncio.Future[str] = asyncio.ensure_future(anext(iterator))
        if cancel_event is None:
            return await chunk_task
        if cancel_event.is_set():
            chunk_task.cancel()
            raise ErpcAbortedError

        async def abort() -> str:
            await cancel_event.wait()
            raise ErpcAbortedError

        cancel_task = asyncio.create_task(abort())
        done, _ = await asyncio.wait({chunk_task, cancel_task}, return_when=asyncio.FIRST_COMPLETED)
        if cancel_task in done:
            chunk_task.cancel()
            await asyncio.gather(chunk_task, return_exceptions=True)
            return await cancel_task
        cancel_task.cancel()
        return await chunk_task

    @staticmethod
    def _update_query(
        ids: Sequence[str],
        encoding: str | None,
        parsed: bool | None,
        ignore_invalid_price_ids: bool | None,
    ) -> Mapping[str, QueryValue]:
        return {
            "ids[]": ids,
            "encoding": encoding,
            "parsed": parsed,
            "ignore_invalid_price_ids": ignore_invalid_price_ids,
        }

    @staticmethod
    def _price_update(value: object) -> PriceUpdateResponse:
        update = _mapping(value)
        binary = _mapping(update.get("binary"))
        data = binary.get("data")
        if (
            not isinstance(data, list)
            or not all(isinstance(item, str) for item in data)
            or not isinstance(binary.get("encoding"), str)
            or ("parsed" in update and not isinstance(update["parsed"], list))
        ):
            raise ErpcInvalidResponseError("ERPC returned an invalid price update")
        return cast(PriceUpdateResponse, update)

    @classmethod
    def _parse_sse_block(cls, block: str) -> PriceStreamEvent | None:
        data: list[str] = []
        result: dict[str, object] = {}
        for line in block.splitlines():
            if not line or line.startswith(":"):
                continue
            field, separator, raw = line.partition(":")
            value = raw[1:] if separator and raw.startswith(" ") else raw
            if field == "data":
                data.append(value)
            elif field in {"event", "id"}:
                result[field] = value
        if not data:
            return None
        try:
            result["data"] = cls._price_update(json.loads("\n".join(data)))
        except (ValueError, ErpcInvalidResponseError) as error:
            raise ErpcInvalidResponseError("ERPC returned malformed stream data") from error
        return cast(PriceStreamEvent, result)


class CloudCatalogClient:
    def __init__(self, transport: RestTransport) -> None:
        self._transport = transport

    async def list(self, options: RequestOptions = DEFAULT_REQUEST_OPTIONS) -> list[CloudOffering]:
        value = await self._transport.get("/v4/cloud/catalog", options=options)
        try:
            envelope = _mapping(value)
            message = _mapping(envelope.get("message"))
            offerings = message.get("offerings")
            if envelope.get("success") is not True or not isinstance(offerings, list):
                raise ErpcInvalidResponseError()
            for raw in offerings:
                item = _mapping(raw)
                for field in ("id", "name", "description"):
                    _string(item.get(field))
                if item.get("kind") not in _RESOURCE_KINDS:
                    raise ErpcInvalidResponseError()
                if item.get("mode") is not None and item.get("mode") not in _RESOURCE_MODES:
                    raise ErpcInvalidResponseError()
                for field in ("regions", "capabilities"):
                    array = item.get(field)
                    if not isinstance(array, list) or not all(
                        isinstance(entry, str) for entry in array
                    ):
                        raise ErpcInvalidResponseError()
        except ErpcInvalidResponseError as error:
            raise ErpcInvalidResponseError("ERPC returned an invalid Cloud catalog") from error
        return cast(list[CloudOffering], offerings)


class CloudCreditClient:
    def __init__(self, transport: RestTransport) -> None:
        self._transport = transport

    async def get(self, options: RequestOptions = DEFAULT_REQUEST_OPTIONS) -> CloudCredit:
        value = await self._transport.get("/v4/cloud/credit", options=options)
        try:
            envelope = _mapping(value)
            if envelope.get("success") is not True:
                raise ErpcInvalidResponseError()
            credit = _mapping(envelope.get("message"))
            if credit.get("alertLevel") not in _ALERT_LEVELS:
                raise ErpcInvalidResponseError()
            _number(credit.get("balanceCents"), integer=True)
            _number(credit.get("burnRateCentsPerHour"), integer=True, nonnegative=True)
            if credit.get("timeToZeroHours") is not None:
                _number(credit.get("timeToZeroHours"), nonnegative=True)
            _datetime(credit.get("quoteTimestamp"))
            _datetime(credit.get("quoteExpiresAt"))
        except ErpcInvalidResponseError as error:
            raise ErpcInvalidResponseError(
                "ERPC returned an invalid Cloud credit snapshot"
            ) from error
        return cast(CloudCredit, credit)


class CloudResourcesClient:
    def __init__(self, transport: RestTransport) -> None:
        self._transport = transport

    async def list(self, options: RequestOptions = DEFAULT_REQUEST_OPTIONS) -> list[CloudResource]:
        value = await self._transport.get("/v4/cloud/resources", options=options)
        try:
            envelope = _mapping(value)
            message = _mapping(envelope.get("message"))
            resources = message.get("resources")
            if envelope.get("success") is not True or not isinstance(resources, list):
                raise ErpcInvalidResponseError()
            for resource in resources:
                self._validate_resource(resource)
        except ErpcInvalidResponseError as error:
            raise ErpcInvalidResponseError(
                "ERPC returned an invalid Cloud resource list"
            ) from error
        return cast(list[CloudResource], resources)

    async def get(
        self, resource_id: str, options: RequestOptions = DEFAULT_REQUEST_OPTIONS
    ) -> CloudResource:
        resource_id = self._resource_id(resource_id)
        value = await self._transport.get(
            f"/v4/cloud/resources/{quote(resource_id, safe='')}", options=options
        )
        try:
            envelope = _mapping(value)
            message = _mapping(envelope.get("message"))
            if envelope.get("success") is not True:
                raise ErpcInvalidResponseError()
            resource = self._validate_resource(message.get("resource"))
        except ErpcInvalidResponseError as error:
            raise ErpcInvalidResponseError("ERPC returned an invalid Cloud resource") from error
        return cast(CloudResource, resource)

    async def get_status(
        self, resource_id: str, options: RequestOptions = DEFAULT_REQUEST_OPTIONS
    ) -> CloudResourceStatus:
        resource_id = self._resource_id(resource_id)
        value = await self._transport.get(
            f"/v4/cloud/resources/{quote(resource_id, safe='')}/status",
            options=options,
        )
        try:
            envelope = _mapping(value)
            if envelope.get("success") is not True:
                raise ErpcInvalidResponseError()
            status = _mapping(envelope.get("message"))
            _string(status.get("id"))
            _string(status.get("status"))
            if "billing" in status:
                billing = _mapping(status["billing"])
                if billing.get("status") not in _BILLING_STATUSES:
                    raise ErpcInvalidResponseError()
                if "hourlyCredits" in billing:
                    _number(billing["hourlyCredits"], nonnegative=True)
                for field in ("nextChargeAt", "graceEndsAt"):
                    if field in billing:
                        _datetime(billing[field])
        except ErpcInvalidResponseError as error:
            raise ErpcInvalidResponseError(
                "ERPC returned an invalid Cloud resource status"
            ) from error
        return cast(CloudResourceStatus, status)

    @staticmethod
    def _resource_id(value: str) -> str:
        result = value.strip()
        if not result:
            raise ErpcConfigError("resource_id must not be empty")
        return result

    @staticmethod
    def _validate_resource(value: object) -> dict[str, object]:
        resource = _mapping(value)
        _string(resource.get("id"))
        _string(resource.get("status"))
        if resource.get("kind") not in _RESOURCE_KINDS:
            raise ErpcInvalidResponseError()
        if resource.get("mode") is not None and resource.get("mode") not in _RESOURCE_MODES:
            raise ErpcInvalidResponseError()
        for field in ("name", "region", "createdAt"):
            if resource.get(field) is not None:
                _string(resource[field])
        return resource

File: packages/ruby/lib/erpc_sdk/rest.rb
# frozen_string_literal: true

require "json"
require "time"

module ERPC
  module Validation
    module_function

    def mapping(value)
      raise InvalidResponseError unless value.is_a?(Hash) && value.keys.all? { |key| key.is_a?(String) }

      value
    end

    def string(value)
      raise InvalidResponseError unless value.is_a?(String)

      value
    end

    def number(value, integer: false, nonnegative: false)
      valid = value.is_a?(Numeric) && value.finite?
      valid &&= value.is_a?(Integer) if integer
      valid &&= value >= 0 if nonnegative
      raise InvalidResponseError unless valid

      value
    end

    def datetime(value)
      text = string(value)
      Time.iso8601(text)
      text
    rescue ArgumentError
      raise InvalidResponseError
    end
  end

  class AccountClient
    def initialize(transport)
      @transport = transport
    end

    def get_token_balance
      data = Validation.mapping(@transport.get("/v3/erpc/token-balance"))
      unless %w[business developer free pro].include?(data["plan"])
        raise InvalidResponseError, "ERPC returned an invalid token balance"
      end
      Validation.number(data["max_tokens"], integer: true)
      Validation.number(data["remaining_tokens"], integer: true)
      Validation.string(data["next_refill_at"]) unless data["next_refill_at"].nil?
      data
    rescue InvalidResponseError
      raise InvalidResponseError, "ERPC returned an invalid token balance"
    end
  end

  class UsageClient
    YEAR_MONTH = /\A\d{4}-(0[1-9]|1[0-2])\z/
    private_constant :YEAR_MONTH

    def initialize(transport)
      @transport = transport
    end

    def get_monthly_api_key_usage(year_month = nil)
      if !year_month.nil? && !YEAR_MONTH.match?(year_month)
        raise ConfigError, "year_month must use YYYY-MM format"
      end

      envelope = Validation.mapping(@transport.get("/v3/user/api-keys/usage", "yearMonth" => year_month))
      raise InvalidResponseError unless envelope["success"] == true

      usage = Validation.mapping(envelope["message"])
      validate_usage(usage)
      usage
    rescue InvalidResponseError
      raise InvalidResponseError, "ERPC returned an invalid monthly API key usage response"
    end

    private

    def validate_usage(usage)
      raise InvalidResponseError unless YEAR_MONTH.match?(Validation.string(usage["yearMonth"]))

      %w[keyCount totalCount totalCredits].each { |field| Validation.number(usage[field]) }
      raise InvalidResponseError unless [true, false].include?(usage["hasStrandedUsage"])
      Validation.string(usage["updatedAt"]) unless usage["updatedAt"].nil?
      validate_chains(usage["chains"])
      raise InvalidResponseError unless usage["apiKeys"].is_a?(Array)

      usage["apiKeys"].each do |raw|
        item = Validation.mapping(raw)
        Validation.string(item["apiKeyLast4"])
        %w[apiKeyLength count credits].each { |field| Validation.number(item[field]) }
        Validation.number(item["keyId"]) unless item["keyId"].nil?
        Validation.string(item["updatedAt"]) unless item["updatedAt"].nil?
        validate_chains(item["chains"])
      end
    end

    def validate_chains(value)
      raise InvalidResponseError unless value.is_a?(Array)

      value.each do |raw_chain|
        chain = Validation.mapping(raw_chain)
        Validation.string(chain["chain"])
        %w[count credits].each { |field| Validation.number(chain[field]) }
        Validation.string(chain["updatedAt"]) unless chain["updatedAt"].nil?
        raise InvalidResponseError unless chain["methods"].is_a?(Array)

        chain["methods"].each do |raw_method|
          method = Validation.mapping(raw_method)
          Validation.string(method["method"])
          %w[count creditCost credits].each { |field| Validation.number(method[field]) }
          Validation.string(method["updatedAt"]) unless method["updatedAt"].nil?
        end
      end
    end
  end

  class PriceClient
    def initialize(transport)
      @transport = transport
    end

    def get_price_feeds(asset_type: nil, query: nil)
      value = @transport.get("/v2/price_feeds", "query" => query, "asset_type" => asset_type)
      raise InvalidResponseError, "ERPC returned invalid price feed metadata" unless value.is_a?(Array)

      value.each do |raw|
        metadata = Validation.mapping(raw)
        Validation.string(metadata["id"])
        attributes = metadata["attributes"]
        next if attributes.nil?
        unless attributes.is_a?(Hash) && attributes.all? { |key, item| key.is_a?(String) && item.is_a?(String) }
          raise InvalidResponseError, "ERPC returned invalid price feed metadata"
        end
      end
      value
    rescue InvalidResponseError
      raise InvalidResponseError, "ERPC returned invalid price feed metadata"
    end

    def get_latest_price_updates(ids:, encoding: nil, parsed: nil, ignore_invalid_price_ids: nil)
      value = @transport.get(
        "/v2/updates/price/latest",
        update_query(ids, encoding, parsed, ignore_invalid_price_ids)
      )
      price_update(value)
    end

    def get_price_updates_at_timestamp(publish_time, ids:, encoding: nil, parsed: nil,
                                       ignore_invalid_price_ids: nil)
      value = @transport.get(
        "/v2/updates/price/#{URLs.escape_path(publish_time)}",
        update_query(ids, encoding, parsed, ignore_invalid_price_ids)
      )
      price_update(value)
    end

    def get_latest_publisher_stake_caps(encoding: nil, parsed: nil)
      Validation.mapping(
        @transport.get(
          "/v2/updates/publisher_stake_caps/latest",
          "encoding" => encoding,
          "parsed" => parsed
        )
      )
    end

    def stream_price_updates(ids:, encoding: nil, parsed: nil, ignore_invalid_price_ids: nil,
                             allow_unordered: nil, benchmarks_only: nil)
      query = update_query(ids, encoding, parsed, ignore_invalid_price_ids).merge(
        "allow_unordered" => allow_unordered,
        "benchmarks_only" => benchmarks_only
      )
      chunks = @transport.stream("/v2/updates/price/stream", query)
      Enumerator.new do |yielder|
        buffer = +""
        chunks.each do |chunk|
          buffer << chunk
          while (boundary = buffer.match(/\r?\n\r?\n/))
            block = buffer.slice!(0, boundary.end(0))
            event = parse_sse_block(block.sub(/\r?\n\r?\n\z/, ""))
            yielder << event if event
          end
        end
        event = parse_sse_block(buffer)
        yielder << event if event
      end
    end

    private

    def update_query(ids, encoding, parsed, ignore_invalid_price_ids)
      {
        "ids[]" => ids,
        "encoding" => encoding,
        "parsed" => parsed,
        "ignore_invalid_price_ids" => ignore_invalid_price_ids
      }
    end

    def price_update(value)
      update = Validation.mapping(value)
      binary = Validation.mapping(update["binary"])
      data = binary["data"]
      valid = data.is_a?(Array) && data.all? { |item| item.is_a?(String) }
      valid &&= binary["encoding"].is_a?(String)
      valid &&= update["parsed"].is_a?(Array) if update.key?("parsed")
      raise InvalidResponseError, "ERPC returned an invalid price update" unless valid

      update
    rescue InvalidResponseError
      raise InvalidResponseError, "ERPC returned an invalid price update"
    end

    def parse_sse_block(block)
      data = []
      result = {}
      block.each_line(chomp: true) do |line|
        next if line.empty? || line.start_with?(":")

        field, separator, raw = line.partition(":")
        value = separator.empty? ? "" : raw.sub(/\A /, "")
        data << value if field == "data"
        result[field] = value if %w[event id].include?(field)
      end
      return nil if data.empty?

      result["data"] = price_update(JSON.parse(data.join("\n")))
      result
    rescue JSON::ParserError, InvalidResponseError
      raise InvalidResponseError, "ERPC returned malformed stream data"
    end
  end

  class CloudCatalogClient
    RESOURCE_KINDS = %w[bare-metal solana-grpc solana-shredstream vps].freeze
    RESOURCE_MODES = %w[dedicated direct shared].freeze

    def initialize(transport)
      @transport = transport
    end

    def list
      envelope = Validation.mapping(@transport.get("/v4/cloud/catalog"))
      message = Validation.mapping(envelope["message"])
      offerings = message["offerings"]
      raise InvalidResponseError unless envelope["success"] == true && offerings.is_a?(Array)

      offerings.each do |raw|
        item = Validation.mapping(raw)
        %w[id name description].each { |field| Validation.string(item[field]) }
        raise InvalidResponseError unless RESOURCE_KINDS.include?(item["kind"])
        raise InvalidResponseError if item["mode"] && !RESOURCE_MODES.include?(item["mode"])
        %w[regions capabilities].each do |field|
          array = item[field]
          raise InvalidResponseError unless array.is_a?(Array) && array.all? { |entry| entry.is_a?(String) }
        end
      end
      offerings
    rescue InvalidResponseError
      raise InvalidResponseError, "ERPC returned an invalid Cloud catalog"
    end
  end

  class CloudCreditClient
    ALERT_LEVELS = %w[critical normal suspended warning].freeze

    def initialize(transport)
      @transport = transport
    end

    def get
      envelope = Validation.mapping(@transport.get("/v4/cloud/credit"))
      raise InvalidResponseError unless envelope["success"] == true

      credit = Validation.mapping(envelope["message"])
      raise InvalidResponseError unless ALERT_LEVELS.include?(credit["alertLevel"])
      Validation.number(credit["balanceCents"], integer: true)
      Validation.number(credit["burnRateCentsPerHour"], integer: true, nonnegative: true)
      Validation.number(credit["timeToZeroHours"], nonnegative: true) unless credit["timeToZeroHours"].nil?
      Validation.datetime(credit["quoteTimestamp"])
      Validation.datetime(credit["quoteExpiresAt"])
      credit
    rescue InvalidResponseError
      raise InvalidResponseError, "ERPC returned an invalid Cloud credit snapshot"
    end
  end

  class CloudResourcesClient
    RESOURCE_KINDS = CloudCatalogClient::RESOURCE_KINDS
    RESOURCE_MODES = CloudCatalogClient::RESOURCE_MODES
    BILLING_STATUSES = %w[active grace-period inactive suspended].freeze

    def initialize(transport)
      @transport = transport
    end

    def list
      envelope = Validation.mapping(@transport.get("/v4/cloud/resources"))
      message = Validation.mapping(envelope["message"])
      resources = message["resources"]
      raise InvalidResponseError unless envelope["success"] == true && resources.is_a?(Array)

      resources.each { |resource| validate_resource(resource) }
      resources
    rescue InvalidResponseError
      raise InvalidResponseError, "ERPC returned an invalid Cloud resource list"
    end

    def get(resource_id)
      value = @transport.get("/v4/cloud/resources/#{URLs.escape_path(normalize_id(resource_id))}")
      envelope = Validation.mapping(value)
      message = Validation.mapping(envelope["message"])
      raise InvalidResponseError unless envelope["success"] == true

      validate_resource(message["resource"])
    rescue InvalidResponseError
      raise InvalidResponseError, "ERPC returned an invalid Cloud resource"
    end

    def get_status(resource_id)
      path = "/v4/cloud/resources/#{URLs.escape_path(normalize_id(resource_id))}/status"
      envelope = Validation.mapping(@transport.get(path))
      raise InvalidResponseError unless envelope["success"] == true

      status = Validation.mapping(envelope["message"])
      Validation.string(status["id"])
      Validation.string(status["status"])
      validate_billing(status["billing"]) if status.key?("billing")
      status
    rescue InvalidResponseError
      raise InvalidResponseError, "ERPC returned an invalid Cloud resource status"
    end

    private

    def normalize_id(value)
      result = value.to_s.strip
      raise ConfigError, "resource_id must not be empty" if result.empty?

      result
    end

    def validate_resource(value)
      resource = Validation.mapping(value)
      Validation.string(resource["id"])
      Validation.string(resource["status"])
      raise InvalidResponseError unless RESOURCE_KINDS.include?(resource["kind"])
      raise InvalidResponseError if resource["mode"] && !RESOURCE_MODES.include?(resource["mode"])
      %w[name region createdAt].each do |field|
        Validation.string(resource[field]) unless resource[field].nil?
      end
      resource
    end

    def validate_billing(value)
      billing = Validation.mapping(value)
      raise InvalidResponseError unless BILLING_STATUSES.include?(billing["status"])
      Validation.number(billing["hourlyCredits"], nonnegative: true) if billing.key?("hourlyCredits")
      %w[nextChargeAt graceEndsAt].each do |field|
        Validation.datetime(billing[field]) if billing.key?(field)
      end
    end
  end
end
